File: src/WordPattern/wordPattern.js
/**
 * @param {string} pattern
 * @param {string} str
 * @return {boolean}
 */
var wordPattern = function(pattern, str) {
    var arr = str.split(' ');
    if (pattern.length != arr.length) return false;
    var smap = new Map(); var pmap = new Map();
    var sa, pa;
    for (var i = 0; i < arr.length; i++) {
        if (smap.has(arr[i])) {
            sa = smap.get(arr[i]);
        }
        if (pmap.has(pattern.charAt(i))) {
            pa = pmap.get(pattern.charAt(i));
        }
        if(sa != pa) return false;
        smap.set(arr[i], i);
        pmap.set(pattern.charAt(i), i);
    }
    return true;
};

File: src/WordPattern/wordPattern.java
class Solution {
    public boolean wordPattern(String pattern, String str) {
        String[] words = str.split(" ");
        if (words.length != pattern.length()) 
            return false;
        Map map = new HashMap();
        for(Integer i = 0; i < words.length; i++) {
            if(map.put(words[i], i) != map.put(pattern.charAt(i) , i)) return false;
        }
        return true;
    }
}